tests/unit/test_context.py

```diff
@@ -0,0 +1,539 @@
+
+
+from fmf.context import Context, InvalidRule, ContextValue, CannotDecide, InvalidContext
+
+
+@pytest.fixture
+def env_centos():
+    return Context(
+        arch="x86_64",
+        distro="centos-8.4.0",
+        component=["bash-5.0.17-1.fc32", "python3-3.8.5-5.fc32"],
+    )
+
+
+class TestExample(object):
+    """ Examples of possible conditions """
+
+    def test_simple_conditions(self, env_centos):
+        """ Rules with singe condition """
+
+        # "Version" comparison if possible
+
+        assert env_centos.matches("distro = centos")
+        assert not env_centos.matches("distro = fedora")
+        assert env_centos.matches("distro = centos,fedora")
+        assert env_centos.matches("distro = centos-8")
+        assert env_centos.matches("distro = centos-8.4")
+
+        assert env_centos.matches("distro > centos-8.3")
+        assert env_centos.matches("distro < centos-9")
+
+        # major/minor operator prefix '~' so it is
+        # False if not comparable (when 'major' differs)
+        with pytest.raises(CannotDecide):
+            env_centos.matches("distro ~< centos-9")
+        with pytest.raises(CannotDecide):
+            env_centos.matches("distro ~>= centos-9")
+        # same major, can be compared
+        assert env_centos.matches("distro ~>= centos-8.2")
+        assert env_centos.matches("distro ~< centos-8.5")
+
+        # Presence of dimension
+        assert env_centos.matches("fips !defined")
+        assert not env_centos.matches("fips defined")
+
+        # operator contains is not necessary anymore
+        assert env_centos.matches("component = bash")
+        assert env_centos.matches("component = bash, ksh")
+        assert not env_centos.matches("component = csh, ksh")
+
+    def test_multi_condition(self, env_centos):
+        """ Rules with multiple conditions """
+
+        assert env_centos.matches("arch=x86_64 && distro != fedora")
+
+        assert env_centos.matches(
+            "distro = fedora && component >= bash-5.0 || distro = centos && component >= bash-4.9"
+        )
+        assert not env_centos.matches(
+            "distro = fedora && component >= bash-5.0 || distro = rhel && component >= bash-4.9"
+        )
+
+    def test_minor_comparisson_mode(self):
+        """ How it minor comparisson should work """
+        fedora = Context(distro="fedora-33")
+        centos = Context(distro="centos-7.3.0")
+        centos6 = Context(distro="centos-6.9.0")
+
+        # Simple version compare is not enough
+        # Think about feature add in centos-7.4.0 and centos-6.9.0
+        # so for 'centos' Context we want matches() == False
+        # but for  'centos6' we want matches() == True
+        rule = "distro >= centos-7.4.0"
+        assert not centos.matches(rule)  # good
+        assert not centos6.matches(rule)  # wrong for us
+
+        rule = "distro >= centos-6.9.0"
+        assert centos.matches(rule)  # wrong for us
+        assert centos6.matches(rule)  # good for us
```

Comment:
What is the meaning of `wrong for us` and `good for us`?

Reply:
Good - Working as we need

Reply:
rewriting to '# we need opposite outcome'